fedimint-client/src/module/init/recovery.rs

```diff
@@ -246,45 +248,53 @@ where
                 .map(move |session_idx| {
                     let api = api.clone();
                     let decoders = decoders.clone();
-                    Box::pin(async move {
-                        info!(session_idx, "Fetching epoch");
-
-                        let mut retry_sleep = Duration::from_millis(10);
-                        let block = loop {
-                            trace!(target: LOG_CLIENT_RECOVERY, session_idx, "Awaiting signed block");
-
-                            let items_res = if core_api_version < VERSION_THAT_INTRODUCED_GET_SESSION_STATUS {
-                                api.await_block(session_idx, &decoders).await.map(|s| s.items)
-                            } else {
-                                api.get_session_status(session_idx, &decoders).await.map(|s| match s {
-                                    SessionStatus::Initial => panic!("Federation missing session that existed when we started recovery"),
-                                    SessionStatus::Pending(items) => items,
-                                    SessionStatus::Complete(s) => s.items,
-                                })
-                            };
+                    let task_group = task_group.clone();
 
-                            match items_res {
-                                Ok(block) => {
-                                    debug!(target: LOG_CLIENT_RECOVERY, session_idx, "Got signed session");
-                                    break block
-                                },
-                                Err(e) => {
-                                    const MAX_SLEEP: Duration = Duration::from_secs(120);
-
-                                    warn!(target: LOG_CLIENT_RECOVERY, e = %e, session_idx, "Error trying to fetch signed block");
-                                    // We don't want PARALLISM_LEVEL tasks hammering Federation
-                                    // with requests, so max sleep is significant
-                                    if retry_sleep <= MAX_SLEEP {
-                                        retry_sleep = retry_sleep
-                                            + thread_rng().gen_range(Duration::ZERO..=retry_sleep);
+                    Box::pin(async move {
+                        // NOTE: Each block is fetched in a spawned task. This avoids a footgun
+                        // of stuff in streams not making any progress when the stream itself
+                        // is not being polled, and possibly can increase the fetching performance.
+                        task_group.spawn_cancellable("recovery fetch block", async move {
```

Comment:
The logic was moved into `spawn_cancellable`.

Reply:
Is this related to the `FuturesUnordered` change or did you just want to move this into a cancellable task?

Reply:
Moving to a task is a solution to FuturesUnordered shortcommings.